README.md
iMTB Report Shiny app
=====================

This is the source code for the Shiny application "iMTB Report". 

The Shiny app relies on the method described in [Perera-Bel et al., 2018](https://doi.org/10.1186/s13073-018-0529-2), and uses the functions available at [MTB-Report repository](https://github.com/jperera-bel/MTB-Report). If you use this method, please cite our publication ([Perera-Bel et al., 2018](https://doi.org/10.1186/s13073-018-0529-2)).

The current release of "iMTB Report" Shiny app uses the latest version of MTB-Report (v1.1.0), which includes the following database versions:

- [Gene Drug Knowledge Database](https://www.synapse.org/#!Synapse:syn2370773): Dienstmann et al., Cancer discovery 5.2 (2015), v20.0
- [CIViC](https://civic.genome.wustl.edu/): Griffith et al., Nat Genet (2017), release 01-Dec-2018
- [TARGET](http://archive.broadinstitute.org/cancer/cga/target): Van Allen et al., Nat Med (2014), v3
- Meric-Bernstam et al., J Natl Cancer Inst. (2015)



Dependencies
------------
```r
install.packages("shiny")
install.packages("shinyjs")
install.packages("DT")
install.packages("shinydashboard")
install.packages("knitr")
install.packages("stringr")
install.packages("xtable")
install.packages("ggplot2")
install.packages("pander")
install.packages("timeSeries")
install.packages("tools")
devtools::install_github("mariodeng/FirebrowseR")
```

Requires LaTeX an Texinfo for the generation of PDF reports. In Linux, install them with:
```
sudo apt-get install texlive-full
sudo apt-get install texinfo
```


Installation
------------

If you are not familiar with R, you can visit our [iMTB Report webpage](http://www.ams.med.uni-goettingen.de:3838/iMTB-Report/app).

Alternatively, you can run it as an stand-alone application in your local computer. Be aware of the [dependencies](#dependencies).
There are many ways to download and run it:

```R
library(shiny)
library(shinyjs)

# Easiest way is to use runGitHub
runGitHub("iMTB-Report", "jperera-bel")

# Run a tar or zip file directly
runUrl("https://github.com/jperera-bel/iMTB-Report/archive/master.tar.gz")
runUrl("https://github.com/jperera-bel/iMTB-Report/archive/master.zip")
```

Or you can clone the git repository, then use `runApp()`:

```R
# First clone the repository with git. If you have cloned it into
# ~/iMTB-Report, first go to that directory, then use runApp().
setwd("~/iMTB-Report")
library(shiny)
library(shinyjs)
runApp()
```






NEWS
______________________________________________
 		v0.2.0
	    Fri 7 Dec, 2018
______________________________________________

Update CIViC; improved layout; filters for results 


DATABASES UPDATES
=================
   - CIViC: Release from 01-Dec-2018

FUNCTIONALITY
=============   
   - MTB-Report (v1.2.0). Updates: 
   	- Created function summary_actionable_genes()
  	- Modified function match_SNV_CIVIC()
   
______________________________________________
 		v0.1.1
	    Fri May 18, 2018
______________________________________________
 
Add tmp/ folder for storing temp files needed for generating PDF reports.
Update README
Update NEWS


______________________________________________
 		v0.1.0
	    Thu May 17, 2018
______________________________________________
 
First release


DATABASES
=========
   - CIViC: Release from 01-May-2018
   - GDKD: v20.0
   -TARGET, v3
   - Meric-Bernstam et al., 2015

FUNCTIONALITY
=============   
   - MTB-Report (v1.1.0)


global.r
library(shiny)
library(DT)
library(shinydashboard)
library(stringr)
library(timeSeries)
library(knitr)
library(ggplot2)   
library(tools)
library(FirebrowseR)
library(shinyjs,lib.loc="/home/jperera/R/x86_64-pc-linux-gnu-library/3.4")

NUM_PAGES <- 5

#######
# DB  #
#######

synonyms = read.csv('data/cancer_types.csv', header = TRUE,sep="\t")
cohorts = read.table('data/tcga_cohorts.csv', header = FALSE,sep=",",stringsAsFactors=F)

##############
# FUNCTIONS  #
##############
myFun <- function(n = 5000) {
  a <- do.call(paste0, replicate(5, sample(LETTERS, n, TRUE), FALSE))
  paste0(a, sprintf("%04d", sample(9999, n, TRUE)), sample(LETTERS, n, TRUE))
}

cosmicLink <- function(gene,var) {
  var2= sub("\\*","",var)
  paste('<a href="http://cancer.sanger.ac.uk/cosmic/search?q=',paste(gene,var2,sep="+AND+"),'" target="_blank">',var,'</a>')
  
}

geneLink  <- function(gene) {
  if (length(gene)==0){return(NULL)}
  paste(paste('<a href="https://pmkb.weill.cornell.edu/search?utf8=✓&search=',gene,'" target="_blank">','PMKB','</a>',sep=""),
  paste('<a href="http://dgidb.genome.wustl.edu/genes/',gene,'" target="_blank">','DGIdb','</a>',sep=""),sep=",")
}

ctLink  <- function(ct) {
  if (length(ct)==0){return(NULL)}
  paste('<a href="https://clinicaltrials.gov/ct2/show/',ct,'" target="_blank">',ct,'</a>',sep="")

}

refLink <- function(ref) {
  if (length(ref)==0){return(links = c())}
  links = c()
  for (x in 1:length(ref)){
    refs=strsplit(ref[x],",")[[1]]
    for (i in 1:length(refs)){
      if (grepl("^\\d",refs[i])){
        links[x] = paste('<a href="https://www.ncbi.nlm.nih.gov/pubmed/?term=',refs[i],'" target="_blank">',refs[i],'</a>')
      }
      if (grepl("ASCO",refs[i])){
        num=strsplit(gsub("[^0-9]","",refs[i]),split = "")[[1]]
        year=num[1:4]
        year=paste(year,collapse = "")
        abstr=num[-(1:4)]
        abstr=paste(abstr,collapse = "")
      #  links[x] = paste('<a href="http://ascopubs.org/doi/abs/10.1200/jco.',year,'.33.15_suppl.',abstr,'" target="_blank">',refs[i],'</a>')
        links[x] = paste('<a href="https://scholar.google.com/scholar?q=%20',paste("ASCO",year,"(abstr",abstr,")",sep="+"),'%20&btnG=&hl=es&as_sdt=0%2C5','" target="_blank">',refs[i],'</a>')
        
      }
      if (grepl("AACR",refs[i])){
        num=strsplit(gsub("AACR| |\\(abstr|\\(abstract|\\)","",refs[i]),split = "")[[1]]
        year=num[1:4]
        year=paste(year,collapse = "")
        abstr=num[-(1:4)]
        abstr=paste(abstr,collapse = "")
       # links[x] = paste('<a href="http://cancerres.aacrjournals.org/search/abstract%252B',abstr,'%20pubyear%3A',year,'%20numresults%3A10%20sort%3Arelevance-rank%20format_result%3Astandard','" target="_blank">',refs[i], '</a>')
        links[x] = paste('<a href="https://scholar.google.com/scholar?q=%20',paste("AACR",year,"(abstr",abstr,")",sep="+"),'%20&btnG=&hl=es&as_sdt=0%2C5','" target="_blank">',refs[i],'</a>')
      }
      if (grepl("FDA",refs[i])){
        links[x] = paste('<a href="https://www.fda.gov/Drugs/ScienceResearch/ResearchAreas/Pharmacogenetics/ucm083378.htm','" target="_blank"> FDA </a>')
      }
      if (!grepl("^\\d|ASCO|AACR|FDA",refs[i])){
        links[x] = paste('<a href="https://scholar.google.com/scholar?q=',gsub(" ","+",refs[i]),'&btnG=&hl=es&as_sdt=0%2C5','" target="_blank">',refs[i],'</a>')
      }
    }
  }
  return(links)
}

helpPopup <- function(title, content,
                      placement=c('right', 'top', 'left', 'bottom'),
                      trigger=c('click', 'hover', 'focus', 'manual')) {
  tagList(
    singleton(
      tags$head(
        tags$script("$(function() { $(\"[data-toggle='popover']\").popover(); })")
      )
    ),
    tags$a(
      href = "#", class = "btn btn-mini", `data-toggle` = "popover",
      title = title, `data-content` = content, `data-animation` = TRUE,
      `data-placement` = match.arg(placement, several.ok=TRUE)[1],
      `data-trigger` = match.arg(trigger, several.ok=TRUE)[1],
      
     # tags$i(class="icon-question-sign")
      icon("info-circle")
    )
  )
}



data/preprocess.R
library(timeSeries)
library(xlsx)

## GDKD: Knowledge_database_v20.0.xlsx" dowloaded from https://www.synapse.org/#!Synapse:syn2370773/files/
## CIViC: 01-May-2018-ClinicalEvidenceSummaries.tsv" dowloaded from https://civic.genome.wustl.edu/releases



	
gdkd         = read.xlsx2('Knowledge_database_v20.0.xlsx', header=TRUE, sheetIndex = 1,colIndex = c(1:45))
civic        = read.delim('01-Dec-2018-ClinicalEvidenceSummaries.tsv', header=T,stringsAsFactors = F,sep="\t",quote = "")


# Small changes
gdkd$Gene  = gsub(" ", "", gdkd$Gene)
for (c in colnames(gdkd)){
	gdkd[,c]  = gsub("^ | $", "", gdkd[,c])
}


for (c in grep("PMID",colnames(gdkd))){
  gdkd[,c] = gsub("\\^|_","",gdkd[,c])
}
civic <- civic[which(civic$evidence_type=="Predictive"),]

# Aggregate databases (join variants with same evidence of: disease, association, drug, evidence level)
gdkd_agg   = aggregate(gdkd, by = list(gdkd$Disease,gdkd$Gene,gdkd$Description,gdkd$Association_1,gdkd$Therapeutic.context_1),
                           FUN = function(X) paste(unique(X), collapse=", "))[,6:50]
civic_agg  = aggregate(civic, by = list(civic$gene,civic$disease,civic$drugs,civic$evidence_level,civic$clinical_significance),
                        FUN = function(X) paste(unique(X), collapse=", "))[,6:41]

gdkd  = gdkd_agg
civic = civic_agg


write.table(gdkd, file="GDKD.csv",col.names=T,row.names=F,sep="\t")
write.table(civic, file="CIViC.csv",col.names=T,row.names=F,sep="\t")




ui.r

my_autocomplete_list <-sort(unique(as.character(synonyms$Shiny)))
tutorial=c("1. Clinical","2. SNVs","3. CNVs","4. Fusions","5. Databases")

sidebar <- dashboardSidebar(
  sidebarMenu(id="tab",
    tags$head(tags$style(HTML('.disclaimer_narrow {padding: 7px;background-color: #34444c ;border-radius: 25px;margin:10px;position:fixed;width:180px;bottom:0px;}
    .main-sidebar .user-panel, .sidebar-menu, .sidebar-menu>li.header {white-space: normal;overflow: hidden;}'))),
    menuItem("Home", tabName = "home",icon = icon('home')),
    menuItem("Tutorial", tabName = "tutorial",icon = icon('info')),
    menuItem("Upload patient data", tabName = "upload", icon = icon('upload')),
    menuItem("Explore TCGA dataset", tabName = "tcga", icon = icon('navicon')),
    HTML("<p class='disclaimer_narrow'>
          <strong>Disclaimer:</strong> This resource is intended for research use only and should not be used for medical or profesional advice. We make no guaran-tee of the comprehensive-ness, reliability or accuracy of the information on this website. You assume full re-sponsibility for all risks as-sociated with using this website.
          </p>" 
         )
  )
)
#shiny-tab-home {width:1200px;padding: 10px 80px;}
body <- dashboardBody(
  tags$head(tags$style(
    HTML('.h1, .h2, .h3, .h4, .h5, .h6, h1, h2, h3, h4, h5, h6 {line-height: 1.5;margin-top: 0px;}
          .h5, h5 {font-size: 14px;}
          .disclaimer {background-color: #f4f4f4;width:700px;border-radius: 25px;padding: 10px;}
          .content-wrapper {background-color: #FFF;}
          .main-header .logo {font-size:19px;}
          .main-sidebar {width: 200px;}
          .navbar { margin-left: 200px; } .content-wrapper, .main-footer, .right-side { margin-left: 200px; }
          .nav-tabs {background-color:#bfbfbf;}
          .nav-tabs-custom .nav-tabs li.active {border-top-color: #009933}
          .nav-tabs-custom .tab-content {padding:  10px 40px 10px 40px;background: #f4f4f4}
          .nav-tabs-custom {border: 1px solid #8c8c8c;}
          .btn-default:hover {background-color: #009933}
          .box {border-top: 3px solid rgba(255,255,255,.15);box-shadow: 0 1px 0px rgba(0, 0, 0, 0)}

          #x1, #x2, #mafUPLOAD, #cnvUPLOAD, #mafHDB, #cnvHDB td {line-height:70%;}
          #clinical {text-align:left;}
          #maf td {padding: 2px 6px;}
          #cnv td {padding: 2px 6px;}
          #actionableSNVs td , #actionableCNVs td {padding: 2px 6px}          
          #resultsTCGA td , #others td , #UPLOAD td , #othersUPLOAD td{padding: 1px 6px;}

          .dataTables_paginate {font-size:9px;}
          .dataTables_length {display: none;}
          .dataTables_wrapper no-footer  {vertical-align: center;}
          .popover-content {
              color: gray;
              font-size: 11px;
          }
        ') #close HTML
       )), #close tags styles
  tabItems(
    tabItem("home",
      fluidRow(
        column(width = 9,
          box(width = 12,
            HTML("
              <center>
              <img src='Logo.jpg' style='width:120px;float: left'>
              <img src='logos.png' style='width:70px;float: right'>
              <h1><strong>Interactive Molecular Tumor Board Report</strong> </h1>
              <h4><strong> - Tool to browse clinically actionable variants of cancer patients -</strong> </h1>
              <h4><i>developed by J&uacutelia Perera-Bel & Tim Bei&szligbarth</h4></i>
              <hr>"),
            box(width = 12,
                h4("Go directly to:"),
                actionButton("upload_button","Upload Patient Data"),
                actionButton("tcga_button","Explore TCGA dataset"),
                br()
                ),
            HTML("
                <h4>1. The user provides somatic variants (<a href='snv_legend.png' target='_blank'>SNVs</a> , <a href='cnv_legend.png' target='_blank'>CNVs</a>, <a href='tx_legend.png' target='_blank'>fusions</a>) of the patient</h4>
                <img src='man_icon.png' style='width:50px'>
                <img src='dna.png' style='width:80px'>
                <img src='upload.png' style='width:60px'>
                <br>
                <br>
                <h4>2. The tool searches in public databases for clinically actionable somatic variants</h4>
                <img src='database_search.png' style='width:60px'>
                <!-- <img src='wordcloud2.png' style='width:250px'> -->
                <img src='man_drug.png' style='width:50px'>
                <br>
                <br>
                <h4>3. The user can browse the patient-specific results in the web interface and also download <a href='Report.pdf' target='_blank'>pdf reports</a></h4>
                <img src='web_browser3.png' style='width:110px'> 
                <img src='download_file2.png' style='width:100px'> 
                </center>" 
                 )
          ) # close box 12
        ), # close column 9
        column(width=3,
          verticalLayout(
            box(title="iMTB Report information",width = 12,status = "primary",solidHeader = T,
              HTML("The <b>Interactive Molecular Tumor Board Report (iMTB Report)</b> is a tool designed to bring together all efforts and knowledge on predictive biomarkers. The tool is oriented at identifying actionable somatic variants of a patient's genomic profile.<br> More detailed information about the method can be found in <a target='_blank' href='https://doi.org/10.1186/s13073-018-0529-2'>our publication (Perera-Bel et al., 2018).</a>")
              ), # close box Info
            box(title="News and Updates",width = 12,status = "primary",solidHeader = T,
             HTML("
             Shiny app source code available in <a target='_blank' href='https://github.com/jperera-bel/iMTB-Report'>GitHub</a>
             <hr>
             Functionality corresponds to latest version of <a target='_blank' href='https://github.com/jperera-bel/MTB-Report'>MTB-Report (v1.2.0)</a>
             <hr>
             Current databases supported:
              <ul>
                <li><a target='_blank' href='https://www.synapse.org/#!Synapse:syn2370773'>Gene Drug Knowledge Database</a>, v20.0</li>
                <li><a target='_blank' href='https://civic.genome.wustl.edu/'>CIViC</a>, 01-Dec-2018</li>
                <li><a target='_blank' href='http://archive.broadinstitute.org/cancer/cga/target'>TARGET</a>, v3</li>
                <li><a target='_blank' href='https://doi.org/10.1093/jnci/djv098'>Meric-Bernstam et al., 2015</a></li>
              </ul>
             ")
            ) # close box News
          ) # close VerticalLayout
        ) # close columns 3
      ) # close Fluidrow
    ), # closes tabItem 1
    tabItem("tutorial",
      "Work in progress"

    ),# closes tabItem tutorial
    tabItem("upload",
      HTML("<h2><strong>Generate MTB-Report of user defined data</strong> </h2>"),
      box(width = 10,
        HTML("<h4>1. First of all, please upload clinical and genomic data of the patient. 
          <br>
          <font color='red'>*</font>
          You have to specify <b>cancer type</b> and upload <b>at least one type of genomic data</b> (<i>SNVs, CNVs or fusions</i>):</h4>"
         )
      ),
     box(width = 12,
      fluidRow(
        tabBox(id="uploadtab",width = 6,side="left",title=strong("Upload Data"),
          tabPanel(title="1. Clinical",
            h5("Cancer type is the only information required by the method, which is used to classify actionable variants into Levels of Evidence. Any other information is merely for completness of the final PDF report."),
            fluidRow(
              column(width=6,
                     selectInput("cancer",label=HTML("Select cancer type<font color='red'>*</font>"),choices=my_autocomplete_list,width = "250px",selected="unspecified"),
                     textInput("ID",label="Type in patient ID",width = "250px"),
                     radioButtons("gender",label="Select gender",choices=c("Female","Male","Other"),width = "250px",selected="none"),
                     selectInput("stage",label="Select cancer stage",choices=c("not specified","I","II","III","IV"),width = "130px")
                     ),
              column(width=6,
                     textInput("prev_ther",label="Type in previous therapies",width = "250px"),
                     radioButtons("tissue",label="Select tissue type",choices=c("Fresh Frozen","FFPE","Other"),width = "250px",selected="none"),
                     textInput("tumor_content",label="Type in tumor content (%)",width = "170px")
                     )
            )
          ),
         tabPanel(title=HTML("2. SNVs"),
            HTML("<b> Upload file with <u>S</u>ingle <u>N</u>ucloetide <u>V</u>ariants (SNVs). Check format <a href='snv_legend.png' target='_blank'>here</a> </b>"),
            br(),
            br(),
            h5("This tool does not check the quality of the variants, so please make sure they are validated and fullfilling your own requirements"),
            fluidRow(
              column(width=5,
               fileInput("inputSNV",label=HTML("<a href='snv.csv' target='_blank'>Click to get example file</a>"),width = "250px"),
               h5(em("File extensions accepted are: .xls, .xlsx, .csv, .dat, .tsv"))
              ),
              column(width=3,
                radioButtons('sep', 'Separator',c(Comma=',',Semicolon=';',Tab='\t'),','),
                checkboxInput(inputId = 'header1', label = 'Header', value = F)
                ),
              column(width=4,HTML('Mandatory columns: Gene Symbol, variant type, aminoacid change')
              ),
             HTML("")
          )),
         tabPanel(title="3. CNVs",
            HTML("<b> Upload file with <u>C</u>opy <u>N</u>umber <u>V</u>ariations (CNVs). Check format <a href='cnv_legend.png' target='_blank'>here</a> </b>"),
            br(),
            br(),
            h5("This tool does not check the quality of the variants, so please make sure they are validated and fullfilling your own requirements"),
            fluidRow(
              column(width=5,
                     fileInput("inputCNV",label=HTML("<a href='cnv.csv' target='_blank'>Click to get example file</a>"),width = "250px"),
                     h5(em("File extensions accepted are: .xls, .xlsx, .csv, .dat, .tsv"))
                     ),
              column(width=3, 
                radioButtons('sep2', 'Separator',c(Comma=',',Semicolon=';', Tab='\t'),','),
                checkboxInput(inputId = 'header2', label = 'Header', value = F)
              ),
              column(width=4,
                     column(width=4,HTML('Mandatory columns: Gene Symbol, variant type')))
          )),
         tabPanel(title="4. Fusions",
            HTML("<b> Upload file with fusions. Check format <a href='tx_legend.png' target='_blank'>here</a> </b>"),
            br(),
            br(),
            h5("This tool does not check the quality of the variants, so please make sure they are validated and fullfilling your own requirements"),
            fluidRow(
              column(width=8,
                     fileInput("inputTX",label=HTML("<a href='fusions.csv' target='_blank'>Click to get example file</a>"),width = "250px"),
                     h5(em("File extensions accepted are: .xls, .xlsx, .csv, .dat, .tsv"))

                     ),
              column(width=4,
                     radioButtons('sep3', 'Separator',c(Comma=',',Semicolon=';',Tab='\t'),','),
                     checkboxInput(inputId = 'header3', label = 'Header', value = F)
                     )
            )),
         tabPanel(title="5. Databases",
            h5(strong(c("Please select the databases used by the report"))),
            br(),
            fluidRow(
              column(width=4,
                     radioButtons('gdkd', em('Gene-Drug Knowledge Database (GDKD)'),
                                  c('v20.0'='default',none='no',other='other'),'default'),
                       conditionalPanel(
                         condition = "input.gdkd == 'other'",
                         fileInput("file_gdkd",label=h6(a(href='https://www.synapse.org/#!Synapse:syn2627707','Please upload a file from repository',target="_blank")),width = "250px")
                       )),
              column(width=4,
                     radioButtons('civic', em('Clinical Interpretations of Variants in Cancer (CIViC)'),
                                  c('01-May-2018'='default',none='no',other='other'),'default'),
                     conditionalPanel(
                                    condition = "input.civic == 'other'",
                                    fileInput("file_civic",label=h6(a(href='https://civicdb.org/releases','Please upload a file from repository',target="_blank")),width = "250px")
                                  ))
              # column(width=4,
              #        radioButtons('cgi', em('BioMarkers from Cancer Genome Interpreter (CGI)'),
              #                     c('last updated 02/08/2017'='default',none=';',other='other'),'default'),
              #        conditionalPanel(
              #                       condition = "input.cgi == 'other'",
              #                       fileInput("file_cgi",label=h6(a(href='https://www.cancergenomeinterpreter.org/biomarkers','Please upload a file from repository',target="_blank")),width = "250px")
              #                     ))
              
            )
         )
       ), # close Upload Data tabBox

       tabBox(width=6,
        side="left",
        title = strong("Browse Genomic Data"),
        tabPanel(title="SNVs ",h5("Single nucleotide variants of the patient's tumor:"),DT::dataTableOutput("mafUPLOAD"),br()),
        tabPanel(title="CNVs ",h5("Copy number variants of the patient's tumor:"),DT::dataTableOutput("cnvUPLOAD"),br()),
        tabPanel(title="Fusions ",h5("Gene fusions of the patient's tumor:"),DT::dataTableOutput("txUPLOAD"),br())
      ) # close Genomic Data tabBox
     ) # close FluidRow
    ), # close box 12

    conditionalPanel(condition=
      "typeof output.mafUPLOAD != 'undefined' || 
      typeof output.cnvUPLOAD != 'undefined' ||
      typeof output.txUPLOAD != 'undefined'",
      box(width = 10,
        HTML(" <h4>2. If the genomic data looks correct, <b>launch the MTB report</b> by clicking the button. 
            <br>The tool will search in public databases for clinically relevant information:</h4>"
            ),
        actionButton("action","Launch/Refresh MTB report",icon = icon("refresh"))
      )
    ), # close conditionalPanel to display ActionButtion
    br(),
    fluidRow(
      conditionalPanel(
        condition = "(input.action > 0 )",
        box(solidHeader = T, width = 12,collapsible = F,
             h3(strong("  Browse clinically relevant genomic data of the selected patient:")),
             HTML("<h4> You can now explore the filtered variants divided into 6 
              <a target='_blank' href='levels.png'>levels of evidence</a>, 
              which determine the actionability of the variant.
              <br>     
              You can also download a .pdf or .csv report with the results. 
              For more information on the method we forward you to our 
              <a target='_blank' href='https://doi.org/10.1186/s13073-018-0529-2'>publication.</a> </h4> "
              ),
             br(),
             box(title = "Summary of Actionable Variants: Filter & Selection",width = 12,collapsible = F,status = 'success',solidHeader = T,
                 column(width=6, 
                   p(strong( "Select Actionable Variants " )) ,
              #     p(strong( "Single Nucleotide Variants " )) ,
                   DT::dataTableOutput('actionableSNVs'),
                   br(),
               #    p(strong( "Copy Number Variants " )) ,
                   DT::dataTableOutput('actionableCNVs')
                  ),
                  column(width=3,    
                    p(strong( "Select Levels of Evidence" )) ,HTML('<font color="gray">Evidence on SAME cancer type</font>'),
                        checkboxGroupInput(
                          "LevelAUPLOAD","",
                          c("A1) FDA & Guidelines"="A1","A2) Clinical Trials"="A2","A3) Pre-clinical"="A3"),selected="none"
                          ),
                        HTML('<font color="gray">Evidence on OTHER cancer types</font>'),
                        checkboxGroupInput(
                          "LevelBUPLOAD","",
                          c("B1) FDA & Guidelines"="B1","B2) Clinical Trials"="B2","B3) Pre-clinical"="B3"),selected="none"
                          ),
                        checkboxInput("checkothersUPLOAD","Display other genes without evidence level",value = FALSE)
                  ),
                 column(width=3,
                  br(),
                  plotOutput(outputId = "figureUPLOAD",height = "170px" , width = "80%"),
                  conditionalPanel(
                          condition=" typeof output.othersUPLOAD != 'undefined' || typeof output.figureUPLOAD != 'undefined'",
                          br(),
                          radioButtons('sortA',"Sort results by:",c("Genes"="genes","Drugs"="drug","Levels of Evidence"="levels"),selected="genes"),
                          h5(strong("Download results")),
                          downloadButton('reportUPLOAD','Download report (.pdf)'),br(),
                          downloadButton('csvUPLOAD','Download report (.csv)')
                        )

                  )
                 ),
             br(),
             box(title = "Details of Actionable Variants",width = 12,collapsible = F,status = 'success',solidHeader = T,
                 column(width=12,
                #  textOutput('cancerGDKD'),
                  # Display Results
                    DT::dataTableOutput('UPLOAD'),
                  # Display Other Results
                    conditionalPanel(
                      condition = "input.checkothersUPLOAD",
                      DT::dataTableOutput('othersUPLOAD')
                    ),
                    br()
                 )
             )
        ) # close box with results
      ) # close conditionalPanel responding to ActionButtion
    ) # close FluidRow
   ), #closes tabItem 2

   tabItem("tcga",
    h2(strong("Explore The Cancer Genome Atlas Dataset")),
    br(),
    box(width = 10,
     HTML(" <h4>1. You can browse up to 34 cancer cohorts. Data is dowloaded using 
      <a href='https://github.com/mariodeng/FirebrowseR' target='_blank'>FirebrowseR</a> package 
      (R client for <a href='http://firebrowse.org/api-docs/' target='_blank'>Broad Firehose Web API</a>)</h4>")
    ),
    box(width = 12,
     fluidRow(
      tabBox(
        title = strong("Select Patient"),
        tabPanel(title="TCGA",
           h5("Please select one cohort from the list below:"),
         # selectInput("cohort","Select cohort",choices = apply(Metadata.Cohorts(format="csv"), 1 , paste , collapse = " - " )),
          selectInput("cohort","Select cohort",choices = paste(cohorts$V1,cohorts$V2, sep = " - " )),
            HTML("<h5>Please select one row from the table and explore its clinically relevant data:</h5>"),
            DT::dataTableOutput('x1')

         )
      ), # close tabBox Select Patient
      conditionalPanel(
       condition = "input.x1_rows_selected.length >0 ",
       tabBox(
         side="left",
         title = strong("Patient Data"),
          tabPanel(title="SNVs ",h5("Single nucleotide variants of the patient's tumor:"),DT::dataTableOutput('maf')),
          tabPanel(title="CNVs ",h5("Copy number variants of the patient's tumor:"),DT::dataTableOutput('cnv')),
          tabPanel(title="Clinical",h5("Main clinical features of the patient's tumor:"),uiOutput('clinical'))
       )
      ) # close ConditionalPanel display genomic
     ) # close FluidRow
    ), # close box 12
    conditionalPanel(
      condition="input.x1_rows_selected.length >0 && (typeof output.maf != 'undefined' || typeof output.cnv != 'undefined')",
      box(width = 10,
        HTML(" <h4>2. The tool will search in public databases for clinically relevant information. To see the results, 
           launch the MTB report by clicking the button. <br>If you select another patient, please remember to click again
           the button to refresh the results. </h4>"),
        actionButton("action2","Launch/Refresh MTB report",icon = icon("refresh"))
      )
    ), # close ConditionalPanel Launch MTB button
    br(),
    br(),
    br(),
    br(),
    fluidRow(
      conditionalPanel(
        condition = "input.x1_rows_selected.length >0 && input.action2 > 0 && (typeof output.maf != 'undefined' || typeof output.cnv != 'undefined')",
        box(solidHeader = T, width = 12,collapsible = F,
             h3(strong("  Browse clinically relevant genomic data of the selected patient:")),
             HTML("<h4> You can now explore the filtered variants divided into 6 
              <a target='_blank' href='levels.png'>levels of evidence</a>, 
              which determine the actionability of the variant.
              <br>     
              You can also download a .pdf or .csv report with the results. 
              For more information on the method we forward you to our 
              <a target='_blank' href='https://doi.org/10.1186/s13073-018-0529-2'>publication.</a> </h4> "
              ),
             br(),
             box(solidHeader = T, width = 12,collapsible = F,
                 column(width=3,                                          
                        checkboxGroupInput(
                          "LevelA","Evidence on SAME cancer type:",
                          c("A1) FDA & Guidelines"="A1","A2) Clinical Trials"="A2","A3) Pre-clinical"="A3"),selected="none"
                          ),
                        checkboxGroupInput(
                          "LevelB","Evidence on OTHER cancer types:",
                          c("B1) FDA & Guidelines"="B1","B2) Clinical Trials"="B2","B3) Pre-clinical"="B3"),selected="none"
                          ),
                        plotOutput(outputId = "figure",height = "170px" , width = "80%"
                          ),
                        br(),
                        checkboxInput("checkothers","Display other genes without evidence level",value = FALSE),
                        conditionalPanel(
                          condition=" typeof output.other != 'undefined' || typeof output.figure != 'undefined'",
                          h5(strong("Download results")),
                          downloadButton('report','Download report (.pdf)'),br(),
                          downloadButton('csv','Download report (.csv)')
                        )
                 ),
                 column(width=9,
                  # Display Results
                    DT::dataTableOutput('resultsTCGA'),
                  # Display Other Results
                    conditionalPanel(
                      condition = "input.checkothers",
                      DT::dataTableOutput('others')
                    )
                 )
             )
        ) # close box with results
      ) # close conditionalPanel responding to ActionButtion
    ) # close FluidRow

   ) #closes tabItem 3


  ) #closes tabItems
) #closes dashboardbody


# Put them together into a dashboardPage
dashboardPage(
  skin = "green",
  dashboardHeader(title="MTB Report",titleWidth = "200px"),
  sidebar,
  body
)
